src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImpl.cs

```diff
@@ -3418,6 +3418,11 @@ private int ReadData()
                         {
                             _ps.bytesUsed = 0;
                         }
+                        else if (bytesLeft < 0)
+                        {
+                            // This can happen when encoding switch causes bytePos to be calculated incorrectly for malformed data
```

Comment:
Would it be better to detect this situation earlier to avoid introducing invalid ParsingState in the first place?

It is hard to reason about what all else can misbehave due to invalid ParsingState.